riscv/src/code_gen_16.rs

```diff
@@ -0,0 +1,1811 @@
+
+    // Sign extends the value in register X and stores it in register Y.
+    // Input is a 32 bit unsigned number. We check bit 7 and set all higher bits to that value.
+    instr sign_extend_byte XL, YL
+        link ~> (tmp1_h, tmp1_l) = regs.mload(XL, STEP)
+        link ~> regs.mstore(YL, STEP + 3, tmp3_h, tmp3_l)
+    {
+        // wrap_bit is used as sign_bit here.
+        tmp1_l = Y_7bit + wrap_bit * 0x80 + X_b2 * 0x100,
+        tmp3_h = wrap_bit * 0xffff,
+        tmp3_l = Y_7bit + wrap_bit * 0xff80
+    }
+
+    col witness Y_7bit;
+    col witness Y_7bit_2;
+    link => bit7.check(Y_7bit);
+    link => bit7.check(Y_7bit_2);
+
+    // Sign extends the value in register X and stores it in register Y.
+    // Input is a 32 bit unsigned number. We check bit 15 and set all higher bits to that value.
+    instr sign_extend_16_bits XL, YL
+        link ~> (tmp1_h, tmp1_l) = regs.mload(XL, STEP)
+        link ~> regs.mstore(YL, STEP + 3, tmp3_h, tmp3_l)
+    {
+        Y_15bit = X_b1 + Y_7bit * 0x100,
+
+        // wrap_bit is used as sign_bit here.
+        tmp1_l = Y_15bit + wrap_bit * 0x8000,
+        tmp3_h = wrap_bit * 0xffff,
+        tmp3_l = tmp1_l
+    }
+    col witness Y_15bit;
+
+    // ======================= assertions =========================
+
+    instr fail 
+        link ~> (tmp1_h, tmp1_l) = regs.mload(0, STEP)
+    {
+      tmp1_h = 1
+    }
+
+    col witness Y_b5;
+    col witness Y_b6;
+    col witness Y_b7;
+    col witness Y_b8;
+    link => byte.check(Y_b5);
+    link => byte.check(Y_b6);
+    link => byte.check(Y_b7);
+    link => byte.check(Y_b8);
+
+    col witness REM_b1;
+    col witness REM_b2;
+    col witness REM_b3;
+    col witness REM_b4;
+    link => byte.check(REM_b1);
+    link => byte.check(REM_b2);
+    link => byte.check(REM_b3);
+    link => byte.check(REM_b4);
+
+    // Computes Q = val(Y) / val(X) and R = val(Y) % val(X) and stores them in registers Z and W.
+    instr divremu YL, XL, ZL, WL
+        link ~> (tmp1_h, tmp1_l) = regs.mload(YL, STEP)
+        link ~> (tmp2_h, tmp2_l) = regs.mload(XL, STEP + 1)
+        link if (1 - XXIsZero) ~> (tmp4_h, tmp4_l, tmp3_h, tmp3_l) = arith_mul.div(tmp1_h, tmp1_l, tmp2_h, tmp2_l)
+        link if (1 - XXIsZero) ~> wrap_bit = arith_bb.cmp(tmp3_h, tmp3_l, tmp2_h, tmp2_l)
+        link if (1 - XXIsZero) ~> wrap_bit_2 = arith_bb.cmp(tmp3_h, tmp3_l, 0, 0)
+        link ~> regs.mstore(ZL, STEP + 2, tmp6_h, tmp6_l)
+        link ~> regs.mstore(WL, STEP + 3, tmp5_h, tmp5_l)
+    {
+        XXIsZero = 1 - XX * XX_inv,
+        XX = tmp2_h + tmp2_l,
+
+        // if X is zero, remainder is set to dividend, as per RISC-V specification:
+        XXIsZero * tmp1_h + (1 - XXIsZero) * tmp3_h = tmp5_h,
+        XXIsZero * tmp1_l + (1 - XXIsZero) * tmp3_l = tmp5_l,
+
+        // remainder >= 0:
+        (1 - XXIsZero) * wrap_bit_2 = 0, 
+
+        // remainder < divisor, conditioned to val(X) not being 0:
+        (1 - XXIsZero) * (wrap_bit - 1) = 0,
```

Comment:
I feel like these checks could also be done in the arith machine, no? I think the division by zero handling is RISC-V specific, but this isn't.

Reply:
fine with me

Reply:
opened https://github.com/powdr-labs/powdr/issues/1891